### PR TITLE
[WIP] variable num_neighbors for different hops & add self loop in the sampled subgraph

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description
Different hops might require different num_neighbors to be sampled (For example, in GraphSage, `num_neigh = 25` for the first layer and `num_neigh = 10` for the second layer). `test_node_flow.py` passed.

TODO: For GraphSage, self loop is needed in the sampled subgraph, which is required to implement an [equivalent version](https://github.com/bkj/pytorch-graphsage/blob/tf-simple/graphsage/supervised_models.py#L140-L176) in DGL.
image

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/mxnet/gcn/gcn_ns.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    def forward(self, nf):
+        nf.layers[0].data['activation'] = nf.layers[0].data['features']
+
+        for i, layer in enumerate(self.layers):
+            h = nf.layers[i].data.pop('activation')
+            if self.dropout:
+                h = mx.nd.Dropout(h, p=self.dropout)
+            nf.layers[i].data['h'] = h
+            nf.flow_compute(fn.copy_src(src='h', out='m'), gcn_reduce, layer, range=i)
```
Comment (zheng-da) [inline examples/mxnet/gcn/gcn_ns.py]: why not use `fn.sum`?

File: examples/mxnet/gcn/gcn_ns.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            self.layers.add(NodeUpdate(n_hidden, n_classes, None))
+
+
+    def forward(self, nf):
+        nf.layers[0].data['activation'] = nf.layers[0].data['features']
+
+        for i, layer in enumerate(self.layers):
+            h = nf.layers[i].data.pop('activation')
+            if self.dropout:
+                h = mx.nd.Dropout(h, p=self.dropout)
```
Comment (zheng-da) [inline examples/mxnet/gcn/gcn_ns.py]: why do you need dropout in the inference?

File: examples/mxnet/gcn/gcn_ns.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            nf.layers[i].data['h'] = h
+            nf.flow_compute(fn.copy_src(src='h', out='m'), gcn_reduce, layer, range=i)
+
+        h = nf.layers[-1].data.pop('activation')
+        return h
+
+
+class GCNInfer(gluon.Block):
```
Comment (zheng-da) [inline examples/mxnet/gcn/gcn_ns.py:66]: what's the difference between GCNInfer and GCNSampling?

File: examples/mxnet/gcn/gcn_ns.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        num_acc = 0.
+
+        for nf, aux in dgl.contrib.sampling.NeighborSampler(g, args.batch_size, g.number_of_nodes(),
+                                                            neighbor_type='in', num_hops=args.n_layers+1,
+                                                            seed_nodes=test_nid):
+            print('test')
+            nf.copy_from_parent()
+            pred = infer_model(nf)
+            batch_nids = nf.layer_parent_nid(-1)
+            num_acc += (pred.argmax(axis=1) == labels[batch_nids]).sum().asscalar()
```
Comment (zheng-da) [inline examples/mxnet/gcn/gcn_ns.py]: This is just for evaluation, maybe we can evaluate on a small batch

**Comment by zheng-da** [comment]
How do you add self-loop in the sampled subgraph?

File: examples/mxnet/gcn/gcn_ns.py (hunk 1 above)
Comment (zheng-da) [reply examples/mxnet/gcn/gcn_ns.py]: if you use built-in functions, it can run much faster.

**Comment by ZiyueHuang** [comment]
I tried skip-connection (second-order proximity from Adaptive Sampling) in `gcn_sc.py`, which has amazing performance. On Cora, it achieves 80% test accuracy after ~30 epochs and finally ~84% test accuracy (if without skip-connection, it achieves 80% test accuracy after ~110 epochs and finally ~81% test accuracy). For neighbor sampling (num_neighbors=2), it stably and quickly achieves ~84% test accuracy (if without skip-connection, it achieves ~60% test accuracy unstably).

```
DGLBACKEND=mxnet python3 gcn_sc.py --dataset cora --num-neighbors NUM_NEIGHBORS  --batch-size 10000 --self-loop

DGLBACKEND=mxnet python3 gcn_ns.py --dataset cora --num-neighbors NUM_NEIGHBORS  --batch-size 10000 --self-loop
```

**Pull request closed without merging**